### PR TITLE
fix: duplicate projects fix and query speed up

### PR DESCRIPTION
**Pull request by keita-determined**
## Description

WEB-802

There is a feature to search workspaces and projects in Web UI. [`GetWorkspaceProjects`](https://github.com/determined-ai/determined/blob/990e62e958d8bf29ac2e58ffb49d2be62ab15a90/master/internal/api_workspace.go#L131) API is called to fetch data. However, it was critically slow to fetch data when there are many workspaces and projects, and returned dupe data.

- Before the fix, it takes around **5 min** to get projects associated with all workspaces
- After the fix, it only takes less than **1 sec**.

I replaced `subquery` with `JOIN` to speed up.

As a result 

|                            | Old Planning Time | New Planning Time | Speed Up Rate |
| --------------- | -------------------: | --------------------: | --------------:  |
| All workspaces | 0.820 ms                  | 7.629 ms                     | -89.251% 📉     |
| A workspace     | 1.063 ms                  | 0.880 ms                    | 120.795% 📈     |

|                            | Old Execution Time | New Execution Time | Speed Up Rate    |
| --------------- | -------------------: | --------------------: | ---------------:  |
| All workspaces | 321442.007 ms        | 2.266 ms                    | 14,185,437% 📈  |
| A workspace     | 105.973 ms               | 0.342 ms                    | 30,986% 📈        |

Overall = Planning Time + Execution Time

|                            | Old Overall Time      | New Overall Time      | Speed Up Rate |
| --------------- | -------------------: | --------------------: | --------------:  |
| All workspaces | 321442.827 ms        | 9.895 ms                    | 3,248,537% 📈 |
| A workspace     | 107.036 ms               | 1.222 ms                    | 8,759% 📈         |

For more details of benchmarks, see [**the comment below**](https://github.com/determined-ai/determined/pull/5744#issuecomment-1382693174)

### Before

https://user-images.githubusercontent.com/109113616/212441124-21457f0b-7ca0-4a44-a3d9-b8ca764c22c2.mov

### After

https://user-images.githubusercontent.com/109113616/212441004-ee5e2ea6-d156-448c-a27c-3871d4433da8.mov

## Test Plan
- Check if API returns the same data as before with the same params  
  - `id=0` means getting all data
  - `id= > 1` means getting a specific workspace id data
- Check if query is fast enough
- Check if data are not duplicated
- Check if it works as before with RBAC enabled

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/static/srv/get_workspace_projects.sql [hunk 1]
```diff
@@ -1,26 +1,38 @@
...
-  UNION SELECT null, null, null
-),
-w AS (
-  SELECT archived
-  FROM workspaces
-  WHERE $1 = 0 OR id = $1
-)
-SELECT p.id, p.name, p.workspace_id, p.description, p.immutable, p.notes,
-  'WORKSPACE_STATE_' || p.state AS state, p.error_message,
+SELECT
```
Comment (keita-determined) [inline master/static/srv/get_workspace_projects.sql:1]: I did `EXPLAIN ANALYZE` to bench the old and new query on local (M1 Max Mac with 32GB memory, 10 cores)

## id=0 (get all data)

### old

```
GroupAggregate  (cost=4761912.07..5090365.51 rows=38628 width=218) (actual time=310136.284..321139.903 rows=262 loops=1)
  Group Key: p.id, u.username, w.archived
  CTE pe
    ->  HashAggregate  (cost=109.51..114.06 rows=455 width=16) (actual time=1.304..1.534 rows=455 loops=1)
          Group Key: experiments.project_id, experiments.state, experiments.start_time
          ->  Append  (cost=0.00..106.10 rows=455 width=16) (actual time=0.014..1.096 rows=455 loops=1)
                ->  Seq Scan on experiments  (cost=0.00..101.54 rows=454 width=16) (actual time=0.014..1.023 rows=454 loops=1)
                ->  Result  (cost=0.00..0.01 rows=1 width=16) (actual time=0.001..0.001 rows=1 loops=1)
  CTE w
    ->  Seq Scan on workspaces  (cost=0.00..6.36 rows=236 width=1) (actual time=0.029..0.771 rows=252 loops=1)
  ->  Sort  (cost=4761791.65..4791589.60 rows=11919180 width=182) (actual time=309876.209..318745.610 rows=15020460 loops=1)
        Sort Key: p.id, u.username, w.archived
        Sort Method: external merge  Disk: 1469936kB
        ->  Hash Left Join  (cost=5.92..183205.84 rows=11919180 width=182) (actual time=1.473..3540.234 rows=15020460 loops=1)
              Hash Cond: (p.user_id = u.id)
              ->  Nested Loop  (cost=0.00..151150.01 rows=11919180 width=148) (actual time=1.369..1947.884 rows=15020460 loops=1)
                    ->  Nested Loop  (cost=0.00..2156.87 rows=107380 width=17) (actual time=1.346..17.889 rows=114660 loops=1)
                          ->  CTE Scan on w  (cost=0.00..4.72 rows=236 width=1) (actual time=0.037..0.993 rows=252 loops=1)
                          ->  CTE Scan on pe  (cost=0.00..9.10 rows=455 width=16) (actual time=0.005..0.034 rows=455 loops=252)
                    ->  Materialize  (cost=0.00..3.67 rows=111 width=131) (actual time=0.000..0.004 rows=131 loops=114660)
                          ->  Seq Scan on projects p  (cost=0.00..3.11 rows=111 width=131) (actual time=0.016..0.066 rows=131 loops=1)
              ->  Hash  (cost=3.74..3.74 rows=174 width=38) (actual time=0.087..0.087 rows=125 loops=1)
                    Buckets: 1024  Batches: 1  Memory Usage: 18kB
                    ->  Seq Scan on users u  (cost=0.00..3.74 rows=174 width=38) (actual time=0.017..0.045 rows=125 loops=1)
Planning time: 0.820 ms
Execution time: 321442.007 ms
```

### New
```
HashAggregate  (cost=136.29..144.23 rows=454 width=218) (actual time=1.873..2.071 rows=131 loops=1)
  Group Key: p.id, w.archived, u.username
  ->  Hash Left Join  (cost=19.72..124.94 rows=454 width=182) (actual time=0.368..1.382 rows=537 loops=1)
        Hash Cond: (p.user_id = u.id)
        ->  Hash Left Join  (cost=13.81..117.80 rows=454 width=148) (actual time=0.260..1.099 rows=537 loops=1)
              Hash Cond: (p.workspace_id = w.id)
              ->  Hash Right Join  (cost=4.50..107.27 rows=454 width=147) (actual time=0.127..0.781 rows=537 loops=1)
                    Hash Cond: (pe.project_id = p.id)
                    ->  Seq Scan on experiments pe  (cost=0.00..101.54 rows=454 width=16) (actual time=0.007..0.203 rows=454 loops=1)
                    ->  Hash  (cost=3.11..3.11 rows=111 width=131) (actual time=0.112..0.112 rows=131 loops=1)
                          Buckets: 1024  Batches: 1  Memory Usage: 21kB
                          ->  Seq Scan on projects p  (cost=0.00..3.11 rows=111 width=131) (actual time=0.014..0.047 rows=131 loops=1)
              ->  Hash  (cost=6.36..6.36 rows=236 width=5) (actual time=0.114..0.115 rows=252 loops=1)
                    Buckets: 1024  Batches: 1  Memory Usage: 18kB
                    ->  Seq Scan on workspaces w  (cost=0.00..6.36 rows=236 width=5) (actual time=0.012..0.059 rows=252 loops=1)
        ->  Hash  (cost=3.74..3.74 rows=174 width=38) (actual time=0.076..0.076 rows=125 loops=1)
              Buckets: 1024  Batches: 1  Memory Usage: 18kB
              ->  Seq Scan on users u  (cost=0.00..3.74 rows=174 width=38) (actual time=0.013..0.031 rows=125 loops=1)
Planning time: 7.629 ms
Execution time: 2.266 ms
```

---

## Id=6 (get data where workspace id is 6)

### Old
```
HashAggregate  (cost=8198.60..8210.78 rows=696 width=218) (actual time=105.617..105.625 rows=4 loops=1)
  Group Key: p.id, u.username, w.archived
  CTE pe
    ->  HashAggregate  (cost=109.51..114.06 rows=455 width=16) (actual time=1.456..1.595 rows=455 loops=1)
          Group Key: experiments.project_id, experiments.state, experiments.start_time
          ->  Append  (cost=0.00..106.10 rows=455 width=16) (actual time=0.023..1.086 rows=455 loops=1)
                ->  Seq Scan on experiments  (cost=0.00..101.54 rows=454 width=16) (actual time=0.022..1.022 rows=454 loops=1)
                ->  Result  (cost=0.00..0.01 rows=1 width=16) (actual time=0.002..0.002 rows=1 loops=1)
  CTE w
    ->  Seq Scan on workspaces  (cost=0.00..6.36 rows=236 width=1) (actual time=0.025..0.122 rows=252 loops=1)
  ->  Nested Loop  (cost=3.54..2709.18 rows=214760 width=182) (actual time=1.607..50.507 rows=229320 loops=1)
        ->  CTE Scan on pe  (cost=0.00..9.10 rows=455 width=16) (actual time=1.463..1.741 rows=455 loops=1)
        ->  Materialize  (cost=3.54..16.76 rows=472 width=166) (actual time=0.000..0.024 rows=504 loops=455)
              ->  Nested Loop  (cost=3.54..14.40 rows=472 width=166) (actual time=0.136..0.910 rows=504 loops=1)
                    ->  CTE Scan on w  (cost=0.00..4.72 rows=236 width=1) (actual time=0.026..0.342 rows=252 loops=1)
                    ->  Materialize  (cost=3.54..3.78 rows=2 width=165) (actual time=0.001..0.001 rows=2 loops=252)
                          ->  Merge Right Join  (cost=3.54..3.77 rows=2 width=165) (actual time=0.105..0.116 rows=2 loops=1)
                                Merge Cond: (u.id = p.user_id)
                                ->  Index Scan using users_pkey on users u  (cost=0.14..17.11 rows=174 width=38) (actual time=0.014..0.021 rows=3 loops=1)
                                ->  Sort  (cost=3.40..3.40 rows=2 width=131) (actual time=0.084..0.085 rows=2 loops=1)
                                      Sort Key: p.user_id
                                      Sort Method: quicksort  Memory: 25kB
                                      ->  Seq Scan on projects p  (cost=0.00..3.39 rows=2 width=131) (actual time=0.023..0.058 rows=2 loops=1)
                                            Filter: (workspace_id = 6)
                                            Rows Removed by Filter: 129
Planning time: 1.063 ms
Execution time: 105.973 ms
```

### New

```
GroupAggregate  (cost=66.17..66.53 rows=8 width=218) (actual time=0.212..0.216 rows=2 loops=1)
  Group Key: p.id, w.archived, u.username
  ->  Sort  (cost=66.17..66.19 rows=8 width=182) (actual time=0.191..0.193 rows=10 loops=1)
        Sort Key: p.id, w.archived, u.username
        Sort Method: quicksort  Memory: 27kB
        ->  Nested Loop Left Join  (cost=3.81..66.05 rows=8 width=182) (actual time=0.100..0.165 rows=10 loops=1)
              ->  Nested Loop Left Join  (cost=3.54..10.76 rows=2 width=166) (actual time=0.086..0.096 rows=2 loops=1)
                    Join Filter: (p.workspace_id = w.id)
                    ->  Merge Right Join  (cost=3.54..3.77 rows=2 width=165) (actual time=0.066..0.075 rows=2 loops=1)
                          Merge Cond: (u.id = p.user_id)
                          ->  Index Scan using users_pkey on users u  (cost=0.14..17.11 rows=174 width=38) (actual time=0.008..0.014 rows=3 loops=1)
                          ->  Sort  (cost=3.40..3.40 rows=2 width=131) (actual time=0.053..0.053 rows=2 loops=1)
                                Sort Key: p.user_id
                                Sort Method: quicksort  Memory: 25kB
                                ->  Seq Scan on projects p  (cost=0.00..3.39 rows=2 width=131) (actual time=0.021..0.037 rows=2 loops=1)
                                      Filter: (workspace_id = 6)
                                      Rows Removed by Filter: 129
                    ->  Materialize  (cost=0.00..6.96 rows=1 width=5) (actual time=0.009..0.009 rows=1 loops=2)
                          ->  Seq Scan on workspaces w  (cost=0.00..6.95 rows=1 width=5) (actual time=0.012..0.012 rows=1 loops=1)
                                Filter: (id = 6)
                                Rows Removed by Filter: 43
              ->  Index Scan using ix_experiments_project_id on experiments pe  (cost=0.27..27.56 rows=9 width=16) (actual time=0.009..0.027 rows=4 loops=2)
                    Index Cond: (p.id = project_id)
Planning time: 0.880 ms
Execution time: 0.342 ms
```

**Comment by keita-determined** [comment]
I replaced `subquery` with `JOIN` because subquery usually makes query slow. In order to benchmark the old and new queries, I did `EXPLAIN ANALYZE` to analyze the queries on local (M1 Max Mac with 32GB memory, 10 cores).
I benchmarked 2 situations with the old and new queries.

1. There are 255 workspaces, 131 projects and 454 experiments in total to get projects associated with all workspaces
   - API `workspaces/0/projects` without params
3. There are 9 experiments, 2 projects in workspace id=6 to get projects associated with a workspace.
   - API `workspaces/6/projects` without params


## id=0 (All workspaces)

<details>
<summary>Old</summary>

```
GroupAggregate  (cost=4761912.07..5090365.51 rows=38628 width=218) (actual time=310136.284..321139.903 rows=262 loops=1)
  Group Key: p.id, u.username, w.archived
  CTE pe
    ->  HashAggregate  (cost=109.51..114.06 rows=455 width=16) (actual time=1.304..1.534 rows=455 loops=1)
          Group Key: experiments.project_id, experiments.state, experiments.start_time
          ->  Append  (cost=0.00..106.10 rows=455 width=16) (actual time=0.014..1.096 rows=455 loops=1)
                ->  Seq Scan on experiments  (cost=0.00..101.54 rows=454 width=16) (actual time=0.014..1.023 rows=454 loops=1)
                ->  Result  (cost=0.00..0.01 rows=1 width=16) (actual time=0.001..0.001 rows=1 loops=1)
  CTE w
    ->  Seq Scan on workspaces  (cost=0.00..6.36 rows=236 width=1) (actual time=0.029..0.771 rows=252 loops=1)
  ->  Sort  (cost=4761791.65..4791589.60 rows=11919180 width=182) (actual time=309876.209..318745.610 rows=15020460 loops=1)
        Sort Key: p.id, u.username, w.archived
        Sort Method: external merge  Disk: 1469936kB
        ->  Hash Left Join  (cost=5.92..183205.84 rows=11919180 width=182) (actual time=1.473..3540.234 rows=15020460 loops=1)
              Hash Cond: (p.user_id = u.id)
              ->  Nested Loop  (cost=0.00..151150.01 rows=11919180 width=148) (actual time=1.369..1947.884 rows=15020460 loops=1)
                    ->  Nested Loop  (cost=0.00..2156.87 rows=107380 width=17) (actual time=1.346..17.889 rows=114660 loops=1)
                          ->  CTE Scan on w  (cost=0.00..4.72 rows=236 width=1) (actual time=0.037..0.993 rows=252 loops=1)
                          ->  CTE Scan on pe  (cost=0.00..9.10 rows=455 width=16) (actual time=0.005..0.034 rows=455 loops=252)
                    ->  Materialize  (cost=0.00..3.67 rows=111 width=131) (actual time=0.000..0.004 rows=131 loops=114660)
                          ->  Seq Scan on projects p  (cost=0.00..3.11 rows=111 width=131) (actual time=0.016..0.066 rows=131 loops=1)
              ->  Hash  (cost=3.74..3.74 rows=174 width=38) (actual time=0.087..0.087 rows=125 loops=1)
                    Buckets: 1024  Batches: 1  Memory Usage: 18kB
                    ->  Seq Scan on users u  (cost=0.00..3.74 rows=174 width=38) (actual time=0.017..0.045 rows=125 loops=1)
Planning time: 0.820 ms
Execution time: 321442.007 ms
```
</details>

<details>
<summary>New</summary>

```
HashAggregate  (cost=136.29..144.23 rows=454 width=218) (actual time=1.873..2.071 rows=131 loops=1)
  Group Key: p.id, w.archived, u.username
  ->  Hash Left Join  (cost=19.72..124.94 rows=454 width=182) (actual time=0.368..1.382 rows=537 loops=1)
        Hash Cond: (p.user_id = u.id)
        ->  Hash Left Join  (cost=13.81..117.80 rows=454 width=148) (actual time=0.260..1.099 rows=537 loops=1)
              Hash Cond: (p.workspace_id = w.id)
              ->  Hash Right Join  (cost=4.50..107.27 rows=454 width=147) (actual time=0.127..0.781 rows=537 loops=1)
                    Hash Cond: (pe.project_id = p.id)
                    ->  Seq Scan on experiments pe  (cost=0.00..101.54 rows=454 width=16) (actual time=0.007..0.203 rows=454 loops=1)
                    ->  Hash  (cost=3.11..3.11 rows=111 width=131) (actual time=0.112..0.112 rows=131 loops=1)
                          Buckets: 1024  Batches: 1  Memory Usage: 21kB
                          ->  Seq Scan on projects p  (cost=0.00..3.11 rows=111 width=131) (actual time=0.014..0.047 rows=131 loops=1)
              ->  Hash  (cost=6.36..6.36 rows=236 width=5) (actual time=0.114..0.115 rows=252 loops=1)
                    Buckets: 1024  Batches: 1  Memory Usage: 18kB
                    ->  Seq Scan on workspaces w  (cost=0.00..6.36 rows=236 width=5) (actual time=0.012..0.059 rows=252 loops=1)
        ->  Hash  (cost=3.74..3.74 rows=174 width=38) (actual time=0.076..0.076 rows=125 loops=1)
              Buckets: 1024  Batches: 1  Memory Usage: 18kB
              ->  Seq Scan on users u  (cost=0.00..3.74 rows=174 width=38) (actual time=0.013..0.031 rows=125 loops=1)
Planning time: 7.629 ms
Execution time: 2.266 ms
```
</details>

🎉🎉 Execution time is **14,185,437% (141,854 times)** faster to get projects associated with all workspaces 🎉🎉

---

## Id=6 (A workspace id = 6)

<details>
<summary>Old</summary>

```
HashAggregate  (cost=8198.60..8210.78 rows=696 width=218) (actual time=105.617..105.625 rows=4 loops=1)
  Group Key: p.id, u.username, w.archived
  CTE pe
    ->  HashAggregate  (cost=109.51..114.06 rows=455 width=16) (actual time=1.456..1.595 rows=455 loops=1)
          Group Key: experiments.project_id, experiments.state, experiments.start_time
          ->  Append  (cost=0.00..106.10 rows=455 width=16) (actual time=0.023..1.086 rows=455 loops=1)
                ->  Seq Scan on experiments  (cost=0.00..101.54 rows=454 width=16) (actual time=0.022..1.022 rows=454 loops=1)
                ->  Result  (cost=0.00..0.01 rows=1 width=16) (actual time=0.002..0.002 rows=1 loops=1)
  CTE w
    ->  Seq Scan on workspaces  (cost=0.00..6.36 rows=236 width=1) (actual time=0.025..0.122 rows=252 loops=1)
  ->  Nested Loop  (cost=3.54..2709.18 rows=214760 width=182) (actual time=1.607..50.507 rows=229320 loops=1)
        ->  CTE Scan on pe  (cost=0.00..9.10 rows=455 width=16) (actual time=1.463..1.741 rows=455 loops=1)
        ->  Materialize  (cost=3.54..16.76 rows=472 width=166) (actual time=0.000..0.024 rows=504 loops=455)
              ->  Nested Loop  (cost=3.54..14.40 rows=472 width=166) (actual time=0.136..0.910 rows=504 loops=1)
                    ->  CTE Scan on w  (cost=0.00..4.72 rows=236 width=1) (actual time=0.026..0.342 rows=252 loops=1)
                    ->  Materialize  (cost=3.54..3.78 rows=2 width=165) (actual time=0.001..0.001 rows=2 loops=252)
                          ->  Merge Right Join  (cost=3.54..3.77 rows=2 width=165) (actual time=0.105..0.116 rows=2 loops=1)
                                Merge Cond: (u.id = p.user_id)
                                ->  Index Scan using users_pkey on users u  (cost=0.14..17.11 rows=174 width=38) (actual time=0.014..0.021 rows=3 loops=1)
                                ->  Sort  (cost=3.40..3.40 rows=2 width=131) (actual time=0.084..0.085 rows=2 loops=1)
                                      Sort Key: p.user_id
                                      Sort Method: quicksort  Memory: 25kB
                                      ->  Seq Scan on projects p  (cost=0.00..3.39 rows=2 width=131) (actual time=0.023..0.058 rows=2 loops=1)
                                            Filter: (workspace_id = 6)
                                            Rows Removed by Filter: 129
Planning time: 1.063 ms
Execution time: 105.973 ms
```
</details>

<details>
<summary>New</summary>

```
GroupAggregate  (cost=66.17..66.53 rows=8 width=218) (actual time=0.212..0.216 rows=2 loops=1)
  Group Key: p.id, w.archived, u.username
  ->  Sort  (cost=66.17..66.19 rows=8 width=182) (actual time=0.191..0.193 rows=10 loops=1)
        Sort Key: p.id, w.archived, u.username
        Sort Method: quicksort  Memory: 27kB
        ->  Nested Loop Left Join  (cost=3.81..66.05 rows=8 width=182) (actual time=0.100..0.165 rows=10 loops=1)
              ->  Nested Loop Left Join  (cost=3.54..10.76 rows=2 width=166) (actual time=0.086..0.096 rows=2 loops=1)
                    Join Filter: (p.workspace_id = w.id)
                    ->  Merge Right Join  (cost=3.54..3.77 rows=2 width=165) (actual time=0.066..0.075 rows=2 loops=1)
                          Merge Cond: (u.id = p.user_id)
                          ->  Index Scan using users_pkey on users u  (cost=0.14..17.11 rows=174 width=38) (actual time=0.008..0.014 rows=3 loops=1)
                          ->  Sort  (cost=3.40..3.40 rows=2 width=131) (actual time=0.053..0.053 rows=2 loops=1)
                                Sort Key: p.user_id
                                Sort Method: quicksort  Memory: 25kB
                                ->  Seq Scan on projects p  (cost=0.00..3.39 rows=2 width=131) (actual time=0.021..0.037 rows=2 loops=1)
                                      Filter: (workspace_id = 6)
                                      Rows Removed by Filter: 129
                    ->  Materialize  (cost=0.00..6.96 rows=1 width=5) (actual time=0.009..0.009 rows=1 loops=2)
                          ->  Seq Scan on workspaces w  (cost=0.00..6.95 rows=1 width=5) (actual time=0.012..0.012 rows=1 loops=1)
                                Filter: (id = 6)
                                Rows Removed by Filter: 43
              ->  Index Scan using ix_experiments_project_id on experiments pe  (cost=0.27..27.56 rows=9 width=16) (actual time=0.009..0.027 rows=4 loops=2)
                    Index Cond: (p.id = project_id)
Planning time: 0.880 ms
Execution time: 0.342 ms
```
</details>

🎉🎉 Execution time is **30,986% (309 times)** faster to get projects associated with a workspace 🎉🎉

## Conclusion

|                            | Old Execution Time | New Execution Time | Speed Up Rate |
| --------------- | -------------------: | --------------------: | --------------:  |
| All workspaces | 321442.007 ms        | 2.266 ms                    | 14,185,437%     |
| a workspace     | 105.973 ms               | 0.342 ms                    | 30,986%           |

**Review by EmilyBonar — APPROVED** [review]
The speedup is great, and I really appreciate the tables and documentation provided. The SQL looks good to me, but you should probably wait on @maxrussell to approve as well.

**Review by maxrussell — APPROVED** [review]
Nice, query looks way more efficient.

Edit: made that comment before seeing your benchmark figures. Pretty hilarious understatement 😂

**Pull request merged**